File: objective-c/APITest/APITest/ZelloAPI.m
//
//  Copyright © 2016 Zello. All rights reserved.
//

#import "ZelloAPI.h"
#import <CommonCrypto/CommonCrypto.h>

/**
 Represents the different HTTP methods that can be used when making requests to the Zello Server API.
 */
typedef NS_ENUM(NSInteger, HTTPMethod) {
  HTTPMethodGET,
  HTTPMethodPOST
};

@interface ZelloAPI ()

@property (nonatomic, strong) NSString *host;
@property (nonatomic, strong) NSString *apiKey;

@end
  
@implementation ZelloAPI

static NSString *version = @"1.1.0";

#pragma mark Version

+ (NSString *)version {
  return version;
}

#pragma mark Initializer

- (id)initWithHost:(NSString *)host apiKey:(NSString *)apiKey {
  return [self initWithHost:host apiKey:apiKey sessionId:NULL];
}

- (id)initWithHost:(NSString *)host apiKey:(NSString *)apiKey sessionId:(NSString *)sessionId {
  self = [super init];
  
  if (self) {
    self.host = host;
    self.apiKey = apiKey;
    self.sessionId = sessionId;
  }
  
  return self;
}

#pragma mark Public Methods

- (void)authenticate:(NSString *)username password:(NSString *)password completionBlock:(ResultCompletionBlock)completionBlock {
  __weak typeof(self) weakSelf = self;
  
  [self callAPI:@"user/gettoken" httpMethod:HTTPMethodGET parameters:nil completionBlock:^(BOOL success, NSDictionary * _Nullable response, NSError * _Nullable error) {
    // On main thread
    if (!weakSelf) {
      completionBlock(NO, response, error);
      return;
    }
    
    if (!success) {
      completionBlock(success, response, error);
      return;
    }
    
    if (!response) {
      completionBlock(NO, response, error);
      return;
    }
    
    NSString *token = [response valueForKey:@"token"];
    
    if (!token) {
      completionBlock(NO, response, error);
      return;
    }
    
    weakSelf.sessionId = [response valueForKey:@"sid"];
    
    if (!weakSelf.apiKey) {
      completionBlock(NO, response, error);
      return;
    }
    
    NSString *hashedPassword = [weakSelf MD5:[[[weakSelf MD5:password] stringByAppendingString:token] stringByAppendingString:weakSelf.apiKey]];
    NSString *parameters = [[[@"username=" stringByAppendingString:username] stringByAppendingString:@"&password="] stringByAppendingString:hashedPassword];
    
    [weakSelf callAPI:@"user/login" httpMethod:HTTPMethodPOST parameters:parameters completionBlock:completionBlock];
  }];
}

- (void)logout:(ResultCompletionBlock)completionBlock {
  __weak typeof(self) weakSelf = self;
  [self callAPI:@"user/logout" httpMethod:HTTPMethodGET parameters:NULL completionBlock:^(BOOL success, NSDictionary * _Nullable response, NSError * _Nullable error) {
    weakSelf.sessionId = NULL;
    
    completionBlock(success, response, error);
  }];
}

- (void)getUsers:(NSString *)username isGateway:(BOOL)isGateway max:(NSNumber *)max start:(NSNumber *)start channel:(NSString *)channel completionBlock:(ResultCompletionBlock)completionBlock {
  NSString *command = @"user/get";
  
  if (username) {
    command = [[command stringByAppendingString:@"/login/"] stringByAppendingString:[self urlEncode:username]];
  }
  if (channel) {
    command = [[command stringByAppendingString:@"/channel/"] stringByAppendingString:[self urlEncode:channel]];
  }
  if (isGateway) {
    command = [command stringByAppendingString:@"/gateway/1"];
  }
  if (max) {
    command = [[command stringByAppendingString:@"/max/"] stringByAppendingString:[max stringValue]];
  }
  if (start) {
    command = [[command stringByAppendingString:@"/start/"] stringByAppendingString:[start stringValue]];
  }
  
  [self callAPI:command httpMethod:HTTPMethodGET parameters:NULL completionBlock:completionBlock];
}

- (void)getChannels:(NSString *)name max:(NSNumber *)max start:(NSNumber *)start completionBlock:(ResultCompletionBlock)completionBlock {
  NSString *command = @"channel/get";
  
  if (name) {
    command = [[command stringByAppendingString:@"/name/"] stringByAppendingString:[self urlEncode:name]];
  }
  if (max) {
    command = [[command stringByAppendingString:@"/max/"] stringByAppendingString:[max stringValue]];
  }
  if (max) {
    command = [[command stringByAppendingString:@"/start/"] stringByAppendingString:[start stringValue]];
  }
  
  [self callAPI:command httpMethod:HTTPMethodGET parameters:NULL completionBlock:completionBlock];
}

- (void)addToChannel:(NSString *)channelName users:(NSArray *)users completionBlock:(ResultCompletionBlock)completionBlock {
  NSString *command = [@"user/addto/" stringByAppendingString:[self urlEncode:channelName]];
  
  NSString *parameters = [self implode:@"login[]=" glue:@"&login[]=" pieces:users];
  
  [self callAPI:command httpMethod:HTTPMethodPOST parameters:parameters completionBlock:completionBlock];
}

- (void)addToChannels:(NSArray *)channelNames users:(NSArray *)users completionBlock:(ResultCompletionBlock)completionBlock {
  NSString *command = @"user/addtochannels";
  
  NSString *parameters = [self implode:@"users[]=" glue:@"&users[]=" pieces:users];
  parameters = [parameters stringByAppendingString:[self implode:@"&channels[]=" glue:@"&channels[]=" pieces:channelNames]];
  
  [self callAPI:command httpMethod:HTTPMethodPOST parameters:parameters completionBlock:completionBlock];
}

- (void)removeFromChannel:(NSString *)channelName users:(NSArray *)users completionBlock:(ResultCompletionBlock)completionBlock {
  NSString *command = [@"user/removefrom/" stringByAppendingString:[self urlEncode:channelName]];
  
  NSString *parameters = [self implode:@"login[]=" glue:@"&login[]=" pieces:users];
  
  [self callAPI:command httpMethod:HTTPMethodPOST parameters:parameters completionBlock:completionBlock];
}

- (void)removeFromChannels:(NSArray *)channelNames users:(NSArray *)users completionBlock:(ResultCompletionBlock)completionBlock {
  NSString *command = @"user/removefromchannels";
  
  NSString *parameters = [self implode:@"users[]=" glue:@"&users[]=" pieces:users];
  parameters = [parameters stringByAppendingString: [self implode:@"&channels[]=" glue:@"&channels[]=" pieces:channelNames]];
  
  [self callAPI:command httpMethod:HTTPMethodPOST parameters:parameters completionBlock:completionBlock];
}

- (void)saveUser:(NSDictionary *)user completionBlock:(ResultCompletionBlock)completionBlock {
  NSString *command = @"user/save";
  
  NSString *parameters = [self createURLStringFromDictionary:user];
  
  [self callAPI:command httpMethod:HTTPMethodPOST parameters:parameters completionBlock:completionBlock];
}

- (void)deleteUsers:(NSArray *)users completionBlock:(ResultCompletionBlock)completionBlock {
  NSString *command = @"user/delete";
  
  NSString *parameters = [self implode:@"login[]=" glue:@"&login[]=" pieces:users];
  
  [self callAPI:command httpMethod:HTTPMethodPOST parameters:parameters completionBlock:completionBlock];
}

- (void)addChannel:(NSString *)name isGroup:(BOOL)isGroup isHidden:(BOOL)isHidden completionBlock:(ResultCompletionBlock)completionBlock {
  NSString *command = [@"channel/add/name/" stringByAppendingString:[self urlEncode:name]];
  
  if (isGroup) {
    command = [[command stringByAppendingString:@"/shared/"] stringByAppendingString:@"true"];
  }
  if (isHidden) {
    command = [[command stringByAppendingString:@"/invisible/"] stringByAppendingString:@"true"];
  }
  
  [self callAPI:command httpMethod:HTTPMethodGET parameters:NULL completionBlock:completionBlock];
}

- (void)deleteChannels:(NSArray *)channelNames completionBlock:(ResultCompletionBlock)completionBlock {
  NSString *command = @"channel/delete";
  
  NSString *parameters = [self implode:@"name[]=" glue:@"&name[]=" pieces:channelNames];
  
  [self callAPI:command httpMethod:HTTPMethodPOST parameters:parameters completionBlock:completionBlock];
}

- (void)getChannelsRoles:(NSString *)channelName completionBlock:(ResultCompletionBlock)completionBlock {
  NSString *command = [@"channel/roleslist/name/" stringByAppendingString:[self urlEncode:channelName]];
  
  [self callAPI:command httpMethod:HTTPMethodGET parameters:NULL completionBlock:completionBlock];
}

- (void)saveChannelRole:(NSString *)channelName roleName:(NSString *)roleName settings:(NSDictionary *)settings completionBlock:(ResultCompletionBlock)completionBlock {
  NSString *command = [[[@"channel/saverole/channel/" stringByAppendingString:[self urlEncode:channelName]] stringByAppendingString:@"/name/"] stringByAppendingString:[self urlEncode:roleName]];
  
  NSString *parameters;
  
  NSError *error;
  NSData *data = [NSJSONSerialization dataWithJSONObject:settings options:kNilOptions error:&error];
  
  NSString *dataString = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
  if (dataString) {
    parameters = [@"settings=" stringByAppendingString:dataString];
  }
  
  [self callAPI:command httpMethod:HTTPMethodPOST parameters:parameters completionBlock:completionBlock];
}

- (void)deleteChannelRole:(NSString *)channelName roles:(NSArray *)roles completionBlock:(ResultCompletionBlock)completionBlock {
  NSString *command = [@"channel/deleterole/channel/" stringByAppendingString:[self urlEncode:channelName]];
  
  NSString *parameters = [self implode:@"roles[]=" glue:@"&roles[]=" pieces:roles];
  
  [self callAPI:command httpMethod:HTTPMethodPOST parameters:parameters completionBlock:completionBlock];
}

- (void)addToChannelRole:(NSString *)channelName roleName:(NSString *)roleName users:(NSArray *)users completionBlock:(ResultCompletionBlock)completionBlock {
  NSString *command = [[[@"channel/addtorole/channel/" stringByAppendingString:[self urlEncode:channelName]] stringByAppendingString:@"/name/"] stringByAppendingString:[self urlEncode:roleName]];
  
  NSString *parameters = [self implode:@"login[]=" glue:@"&login[]=" pieces:users];
  
  [self callAPI:command httpMethod:HTTPMethodPOST parameters:parameters completionBlock:completionBlock];
}

#pragma mark Private Methods

- (void)callAPI:(NSString *)command httpMethod:(HTTPMethod)method parameters:(NSString *)parameters completionBlock:(ResultCompletionBlock)completionBlock {
  NSURLSession *session = [NSURLSession sharedSession];
  
  __weak typeof(self) weakSelf = self;

  if (!self.host) {
    dispatch_async(dispatch_get_main_queue(), ^{
      completionBlock(NO, NULL, [weakSelf unknownError]);
    });
    return;
  }
  
  NSString *prefix = @"http://";
  if ([self.host containsString:@"http://"] || [self.host containsString:@"https://"]) {
    prefix = @"";
  }
  
  NSString *urlString = [prefix stringByAppendingString:[[self.host stringByAppendingString:@"/"] stringByAppendingString:command]];
  
  if (self.sessionId) {
    urlString = [[urlString stringByAppendingString:@"?sid="] stringByAppendingString:_sessionId];
  }
  
  self.lastURL = urlString;
  
  NSURL *url = [[NSURL alloc] initWithString:urlString];
  
  if (!url) {
    dispatch_async(dispatch_get_main_queue(), ^{
      completionBlock(NO, NULL, [weakSelf unknownError]);
    });
    return;
  }
  
  NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL:url];
  request.HTTPMethod = [self convertToString:method];
  
  if (parameters) {
    request.HTTPBody = [parameters dataUsingEncoding:NSUTF8StringEncoding];
    
    [request addValue:@"application/x-www-form-urlencoded; charset=utf-8" forHTTPHeaderField:@"Content-Type"];
  }
  
  [[session dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
    if (!error) {
      if (!data) {
        dispatch_async(dispatch_get_main_queue(), ^{
          completionBlock(NO, NULL, [weakSelf unknownError]);
        });
        return;
      }
      
      NSError *error;
      NSDictionary *responseDictionary = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:&error];
      
      if (error) {
        dispatch_async(dispatch_get_main_queue(), ^{
          completionBlock(NO, NULL, error);
        });
        return;
      }
      
      if (responseDictionary) {
        NSString *statusCode = [responseDictionary valueForKey:@"code"];
        if (statusCode) {
          // Happy path
          BOOL success = [statusCode isEqualToString:@"200"];
          dispatch_async(dispatch_get_main_queue(), ^{
            completionBlock(success, responseDictionary, NULL);
          });
          return;
        } else {
          dispatch_async(dispatch_get_main_queue(), ^{
            completionBlock(NO, NULL, [weakSelf unknownError]);
          });
          return;
        }
      } else {
        dispatch_async(dispatch_get_main_queue(), ^{
          completionBlock(NO, NULL, [weakSelf unknownError]);
        });
        return;
      }
    } else {
      dispatch_async(dispatch_get_main_queue(), ^{
        completionBlock(NO, NULL, error);
      });
      return;
    }
  }] resume];
}

/**
 *  Converts HTTPMethod to NSString.
 *
 *  @param method method to convert.
 *
 *  @return NSString representation of method.
 */
- (NSString *)convertToString:(HTTPMethod)method {
  NSString *result = nil;
  
  switch(method) {
    case HTTPMethodGET:
      result = @"GET";
      break;
    case HTTPMethodPOST:
      result = @"POST";
      break;
  }
  
  return result;
}

/**
 Calculates the MD5 hash of a string.
 */
- (NSString *)MD5:(NSString *)string {
  const char *pointer = [string UTF8String];
  unsigned char md5Buffer[CC_MD5_DIGEST_LENGTH];
  
  CC_MD5(pointer, (CC_LONG)strlen(pointer), md5Buffer);
  
  NSMutableString *returnString = [NSMutableString stringWithCapacity:CC_MD5_DIGEST_LENGTH * 2];
  for (int i = 0; i < CC_MD5_DIGEST_LENGTH; i++) {
    [returnString appendFormat:@"%02x",md5Buffer[i]];
  }
  
  return returnString;
}

/**
 URL encodes a string.
 */
- (NSString *)urlEncode:(NSString *)string {
  NSString *returnString = [string stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet alphanumericCharacterSet]];
  
  return returnString ? returnString : string;
}

/**
 Join array elements with a glue string. URL encodes all pieces.
 */
- (NSString *)implode:(NSString *)baseString glue:(NSString *)glue pieces:(NSArray *)pieces {
  for (int i = 0; i < pieces.count; i++) {
    baseString = [baseString stringByAppendingString:[self urlEncode:pieces[i]]];
    
    if (i < pieces.count - 1) {
      baseString = [baseString stringByAppendingString:glue];
    }
  }
  
  return baseString;
}

- (NSError *)unknownError {
  return [[NSError alloc] initWithDomain:@"ZELLOUnknownError" code:0 userInfo:nil];
}

- (NSString *)createURLStringFromDictionary:(NSDictionary *)dictionary {
  NSString *string = @"";
  for (int i = 0; i < dictionary.allKeys.count; i++) {
    NSString *key = dictionary.allKeys[i];
    NSString *value = [dictionary valueForKey:key];
    
    if (i == 0) {
      string = [[[string stringByAppendingString:key] stringByAppendingString:@"="] stringByAppendingString:[self urlEncode:value]];
    } else {
      string = [[[[string stringByAppendingString:@"&"] stringByAppendingString:key] stringByAppendingString:@"="] stringByAppendingString:[self urlEncode:value]];
    }
  }
  
  return string;
}

@end

File: swift/ZelloAPI.swift
//
//  Copyright © 2016 Zello. All rights reserved.
//

import Foundation

/// Completion Handler for making API Requests that returns a success indicator, result dictionary and error.
/// A value for the error variable represents a client error. This is not the error returned by the API.
/// An error returned by the API can be retrieved through the result dictionary.
/// This closure is always executed on the main thread.
public typealias ResultCompletionHandler = (Bool, [String : AnyObject]?, NSError?) -> Void

/**
 Zello for Work server Swift API wrapper class.

 The class provides an easy way to interact with Zello for Work server
 from your Swift code to add, modify and delete users and channels.
 Please note that all text values passed to the API must be in UTF-8 encoding
 and any text data returned are in UTF-8 as well.

 - Version 1.0.0
 - Swift Version 2.2
 - Minimum iOS Version 7.0
*/
public class ZelloAPI {
  
  // MARK: Public Variables
  
  /// API Version
  public static let version = "1.0.0"
  
  /// Session ID used to identify logged in client. Typically you'll want to authenticate first and store the Session ID to reuse later.
  public var sessionId: String?
  
  /// Last accessed API URL. Useful for API troubleshooting.
  public var lastURL: String?
  
  // MARK: Private Variables
  
  /// Server hostname or IP address
  private var host: String?
  /// API Key
  private var apiKey: String?
  
  // MARK: Initializer
  
  public convenience init(host: String, apiKey: String, sessionId: String?=nil) {
    self.init()
    
    self.host = host
    self.apiKey = apiKey
    self.sessionId = sessionId
  }
  
  // MARK: Public Methods
  
  /**
   API client authentication.
   If authentication fails, use the errorCode and errorDescription attributes on the response dictionary to get error details.
   If authentication succeeds, sessionId is set to the Session ID.
   The Session ID is reusable so it's recommended that you save this value and use it for further API calls.
   Once you are done using the API, call ZelloAPI.logout() to end the session and invalidate Session ID.
   
   - parameter username:          administrative username
   - parameter password:          administrative password
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  public func authenticate(username: String, password: String, completionHandler: ResultCompletionHandler) {
    callAPI("user/gettoken", httpMethod: .GET, completionHandler: { [weak self] (success, response, error) -> Void in
      // On main thread
      guard let weakSelf = self else {
        completionHandler(false, response, error)
        return
      }
      
      if !success {
        completionHandler(success, response, error)
        return
      }
      
      guard let response = response else {
        completionHandler(false, nil, error)
        return
      }
      
      guard let token = response["token"] as? String else {
        completionHandler(false, response, error)
        return
      }
      
      weakSelf.sessionId = response["sid"] as? String
      
      guard let apiKey = weakSelf.apiKey else {
        completionHandler(false, response, error)
        return
      }
      
      let parameters = "username=\(username)&password=\((password.MD5() + token + apiKey).MD5())"
      
      weakSelf.callAPI("user/login", httpMethod: .POST, parameters: parameters, completionHandler: completionHandler)
    })
  }
  
  /**
   Ends session identified by sessionId.
   Use this method to terminate the API session.
   See ZelloAPI.authenticate()
   
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  public func logout(completionHandler: ResultCompletionHandler) {
    callAPI("user/logout", httpMethod: .GET, completionHandler: { [weak self] (success, response, error) -> Void in
      self?.sessionId = nil
      
      completionHandler(success, response, error)
    })
  }
  
  /**
   Gets the list of the users or detailed informatino regarding a particular user.
   
   - parameter username:          username of the user, for which the details are requested. If nil, the full users list is returned.
   - parameter isGateway:         whether to return users or gateways.
   - parameter max:               maximum number of results to fetch.
   - parameter start:             start index of results to fetch.
   - parameter channel:           channel name.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  public func getUsers(username: String?=nil, isGateway: Bool?=false, max: Int?=nil, start: Int?=nil, channel: String?=nil, completionHandler: ResultCompletionHandler) {
    var command = "user/get"
    
    if let username = username {
      command += "/login/" + username.urlEncode()
    }
    if let channel = channel {
      command += "/channel/" + channel.urlEncode()
    }
    if let isGateway = isGateway where isGateway {
      command += "/gateway/1"
    }
    if let max = max {
      command += "/max/" + String(max)
    }
    if let start = start {
      command += "/start/" + String(start)
    }
    
    callAPI(command, httpMethod: .GET, completionHandler: completionHandler)
  }
  
  /**
   Gets the list of the channels or detailed information regarding a particular channel.
   
   - parameter name:              name of the channel, for which the details are requested. If omitted, the full channels list is returned.
   - parameter max:               maximum number of results to fetch.
   - parameter start:             start index of results to fetch.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  public func getChannels(name: String?=nil, max: Int?=nil, start: Int?=nil, completionHandler: ResultCompletionHandler) {
    var command = "channel/get"
    
    if let name = name {
      command += "/name/" + name.urlEncode()
    }
    if let max = max {
      command += "/max/" + String(max)
    }
    if let start = start {
      command += "/start/" + String(start)
    }
    
    callAPI(command, httpMethod: .GET, completionHandler: completionHandler)
  }
  
  /**
   Adds users to a channel.
   
   - parameter channelName: name of the channel where the users will be added.
   - parameter users:       usernames of the users to add.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  public func addToChannel(channelName: String, users: [String], completionHandler: ResultCompletionHandler) {
    let command = "user/addto/" + channelName.urlEncode()
    
    let parameters = "login[]=".implode("&login[]=", pieces: users)

    callAPI(command, httpMethod: .POST, parameters: parameters, completionHandler: completionHandler)
  }
  
  /**
   Add users to multiple channels.
   
   - parameter channelNames:      channel names where the users are added.
   - parameter users:             usernames of the users to add.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  public func addToChannels(channelNames: [String], users: [String], completionHandler: ResultCompletionHandler) {
    let command = "user/addtochannels"
    
    var parameters = "users[]=".implode("&users[]=", pieces: users)
    parameters += "&channels[]=".implode("&channels[]=", pieces: channelNames)
    
    callAPI(command, httpMethod: .POST, parameters: parameters, completionHandler: completionHandler)
  }
  
  /**
   Removes users from a channel.
   
   - parameter channelName:       name of the channel.
   - parameter users:             usernames of the users to remove from the channel.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  public func removeFromChannel(channelName: String, users: [String], completionHandler: ResultCompletionHandler) {
    let command = "user/removefrom/" + channelName.urlEncode()
    
    let parameters = "login[]=".implode("&login[]=", pieces: users)
    
    callAPI(command, httpMethod: .POST, parameters: parameters, completionHandler: completionHandler)
  }
  
  /**
   Removes users from multiple channels.
   
   - parameter channelNames:      names of the channels.
   - parameter users:             usernames of the users to remove from channels.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  public func removeFromChannels(channelNames: [String], users: [String], completionHandler: ResultCompletionHandler) {
    let command = "user/removefromchannels"
    
    var parameters = "users[]=".implode("&users[]=", pieces: users)
    parameters += "&channels[]=".implode("&channels[]=", pieces: channelNames)
    
    callAPI(command, httpMethod: .POST, parameters: parameters, completionHandler: completionHandler)
  }
  
  /**
   Adds or updates the user.
   If username exists, the user is updated. Otherwise, a new user is created.
   When adding a user, the "name" and "password" attributes are required.
   When updating a user, the "name" attribute is required.
   
   See ZelloAPI.deleteUsers()
   
   Attributes:
   - name (required) - username
   - password - password md5 hash
   - email - e-mail address
   - full_name - user alias
   - job - user position
   - admin - "true" or "false". Defines whether the user has access to the admin console.
   - limited_access - "true" or "false". Defines whether the user is restricted from starting 1-on-1 conversations or not.
   - gateway - set to "true" for adding a gateway, "false" if normal user.
   - add - "true" or "false". If set to "true", the existing user will not be updated and an error will be returned.
   
   - parameter user:              dictionary of user attributes.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  public func saveUser(user: [String : String], completionHandler: ResultCompletionHandler) {
    let command = "user/save"
    
    let parameters = createURLStringFromDictionary(user)
    
    callAPI(command, httpMethod: .POST, parameters: parameters, completionHandler: completionHandler)
  }
  
  /**
   Deletes users.
   
   See ZelloAPI.saveUser()
   
   - parameter users:             usernames of the users to remove.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  public func deleteUsers(users: [String], completionHandler: ResultCompletionHandler) {
    let command = "user/delete"
    
    let parameters = "login[]=".implode("&login[]=", pieces: users)
    
    callAPI(command, httpMethod: .POST, parameters: parameters, completionHandler: completionHandler)
  }
  
  /**
   Adds a new channel.
   
   - parameter name:              channel name.
   - parameter isGroup:           true means it is a group channel. false means it is a dynamic channel.
   - parameter isHidden:          when set to true in combination with isGroup set to true, a hidden group channel is created.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  public func addChannel(name: String, isGroup: Bool?=nil, isHidden: Bool?=nil, completionHandler: ResultCompletionHandler) {
    var command = "channel/add/name/" + name.urlEncode()
    
    if let isGroup = isGroup {
      command += "/shared/" + (isGroup ? "true" : "false")
    }
    if let isHidden = isHidden {
      command += "/invisible/" + (isHidden ? "true" : "false")
    }
    
    callAPI(command, httpMethod: .GET, completionHandler: completionHandler)
  }
  
  /**
   Deletes channels.
   
   - parameter channelNames:      names of the channels to remove.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  public func deleteChannels(channelNames: [String], completionHandler: ResultCompletionHandler) {
    let command = "channel/delete"
    
    let parameters = "name[]=".implode("&name[]=", pieces: channelNames)
    
    callAPI(command, httpMethod: .POST, parameters: parameters, completionHandler: completionHandler)
  }
  
  /**
   Get channel roles (simple format).
   
   - parameter channelName:       channel name.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  public func getChannelsRoles(channelName: String, completionHandler: ResultCompletionHandler) {
    let command = "channel/roleslist/name/" + channelName.urlEncode()
    
    callAPI(command, httpMethod: .GET, completionHandler: completionHandler)
  }
  
  /**
   Adds or updates channel role.
   
   - parameter channelName:       channel name.
   - parameter roleName:          new role name.
   - parameter settings:          role settings in JSON format: ["listen_only" : false, "no_disconnect" : true, "allow_alerts" : false, "to": ["dispatchers"]]
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  public func saveChannelRole(channelName: String, roleName: String, settings: [String : AnyObject], completionHandler: ResultCompletionHandler) {
    let command = "channel/saverole/channel/" + channelName.urlEncode() + "/name/" + roleName.urlEncode()
    
    var parameters: String?
    do {
      let data = try NSJSONSerialization.dataWithJSONObject(settings, options: NSJSONWritingOptions(rawValue: 0))
      if let dataString = String(data: data, encoding: NSUTF8StringEncoding) {
        parameters = "settings=" + dataString
      }
    } catch {
      
    }
    
    callAPI(command, httpMethod: .POST, parameters: parameters, completionHandler: completionHandler)
  }
  
  /**
   Deletes channel role.
   
   - parameter channelName:       channel name.
   - parameter roles:             role names to delete.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  public func deleteChannelRole(channelName: String, roles: [String], completionHandler: ResultCompletionHandler) {
    let command = "channel/deleterole/channel/" + channelName.urlEncode()
    
    let parameters = "roles[]=".implode("&roles[]=", pieces: roles)
    
    callAPI(command, httpMethod: .POST, parameters: parameters, completionHandler: completionHandler)
  }
  
  /**
   Adds users to role in channel.
   
   - parameter channelName:       channel name.
   - parameter roleName:          role name.
   - parameter users:             usernames to add to role in channel.
   - parameter completionHandler: completion handler indicating success, response and error.
   */
  public func addToChannelRole(channelName: String, roleName: String, users: [String], completionHandler: ResultCompletionHandler) {
    let command = "channel/addtorole/channel/" + channelName.urlEncode() + "/name/" + roleName.urlEncode()
    
    let parameters = "login[]=".implode("&login[]=", pieces: users)
    
    callAPI(command, httpMethod: .POST, parameters: parameters, completionHandler: completionHandler)
  }
  
  // MARK: Private Methods
  
  private func callAPI(command: String, httpMethod: HTTPMethod, parameters: String?=nil, completionHandler: ResultCompletionHandler) {
    let session = NSURLSession.sharedSession()
    
    guard let host = host else {
      return
    }
    
    var urlString = host + "/" + command

    if let sessionId = sessionId {
      urlString += "?sid=" + sessionId
    }
    
    lastURL = urlString
    
    guard let nsURL = NSURL(string: urlString) else {
      dispatch_async(dispatch_get_main_queue(), { 
        completionHandler(false, nil, NSError.unknownError())
      })
      return
    }
    
    let request = NSMutableURLRequest(URL: nsURL)
    request.HTTPMethod = httpMethod.rawValue
    
    if let parameters = parameters {
      request.setBodyContentFromString(parameters)
    }
    
    session.dataTaskWithRequest(request) { (data, response, error) -> Void in
      if error == nil {
        guard let data = data else {
          dispatch_async(dispatch_get_main_queue(), { 
            completionHandler(false, nil, NSError.unknownError())
          })
          return
        }
        
        var responseDictionary: [String : AnyObject]?
        do {
          responseDictionary = try NSJSONSerialization.JSONObjectWithData(data, options: NSJSONReadingOptions(rawValue: 0)) as? [String : AnyObject]
        } catch let error as NSError {
          dispatch_async(dispatch_get_main_queue(), { 
            completionHandler(false, nil, error)
          })
          return
        }
        
        guard let unwrappedResponseDictionary = responseDictionary else {
          dispatch_async(dispatch_get_main_queue(), {
            completionHandler(false, nil, NSError.unknownError())
          })
          return
        }
        
        guard let statusCode = unwrappedResponseDictionary["code"] as? String else {
          dispatch_async(dispatch_get_main_queue(), {
            completionHandler(false, unwrappedResponseDictionary, NSError.unknownError())
          })
          return
        }
        
        let success = statusCode == "200"
        dispatch_async(dispatch_get_main_queue(), { 
          completionHandler(success, unwrappedResponseDictionary, nil)
        })
        return
      } else {
        dispatch_async(dispatch_get_main_queue(), { 
          completionHandler(false, nil, NSError.unknownError())
        })
        return
      }
    }.resume()
  }
  
  private func createURLStringFromDictionary(dictionary: [String : String]) -> String {
    var string = ""
    for (index, key) in dictionary.keys.enumerate() {
      if index == 0 {
        string += key + "=" + dictionary[key]!.urlEncode()
      } else {
        string += "&" + key + "=" + dictionary[key]!.urlEncode()
      }
    }
    
    return string
  }

}

// MARK: Helpers

/**
 Represents the different HTTP methods that can be used when making requests to the Zello Server API.
 */
private enum HTTPMethod: String {
  
  case POST = "POST"
  case GET = "GET"
  
}

private enum ZelloAPIErrorDomain: String {
  
  case UnknownError = "ZELLOUnknownError"

}

private extension NSError {
  
  static func unknownError() -> NSError {
    return NSError(domain: ZelloAPIErrorDomain.UnknownError.rawValue, code: 0, userInfo: nil)
  }
  
}

private extension String {
  
  /// Calculates the MD5 hash of a string.
  func MD5() -> String {
    guard let str = cStringUsingEncoding(NSUTF8StringEncoding) else {
      return ""
    }
    
    let strLen = CC_LONG(lengthOfBytesUsingEncoding(NSUTF8StringEncoding))
    let digestLen = Int(CC_MD5_DIGEST_LENGTH)
    let result = UnsafeMutablePointer<CUnsignedChar>.alloc(digestLen)
    
    CC_MD5(str, strLen, result)
    
    let hash = NSMutableString()
    for i in 0..<digestLen {
      hash.appendFormat("%02x", result[i])
    }
    
    result.dealloc(digestLen)
    
    return String(format: hash as String)
  }
  
  /// URL encodes a string.
  func urlEncode() -> String {
    if let string = stringByAddingPercentEncodingWithAllowedCharacters(NSCharacterSet.alphanumericCharacterSet()) {
      return string
    }
    
    return self
  }

  /**
   Join array elements with a glue string. URL encodes all pieces.
   */
  func implode(glue: String, pieces: [String]) -> String {
    var string = self
    
    for (index, piece) in pieces.enumerate() {
      string += piece.urlEncode()
      
      if (index < pieces.count - 1) {
        string += glue
      }
    }
    
    return string
  }
  
}

private extension NSMutableURLRequest {
  
  /**
   Sets the HTTPBody of the NSMutableURLRequest to the contents of a string.
   
   - parameter string: string to encode into HTTPBody.
   */
  func setBodyContentFromString(string: String) {
    HTTPBody = string.dataUsingEncoding(NSUTF8StringEncoding)
    
    addValue("application/x-www-form-urlencoded; charset=utf-8", forHTTPHeaderField: "Content-Type")
  }
  
}
